packages/gatsby/src/internal-plugins/query-runner/query-queue.js

```diff
@@ -1,106 +1,109 @@
-  filter: (job, cb) => {
-    if (processing.has(job.id)) {
-      waiting.set(job.id, job)
-      cb(`already running`)
-    } else {
-      cb(null, job)
-    }
-  },
-  store: FastMemoryStore(),
+const makeBaseOptions = () => {
+  return {
+    concurrent: 4,
+    store: FastMemoryStore(),
+  }
 }
 
-// During builds we don't need all the filtering, etc. so we
-// remove them to speed up queries
-if (process.env.gatsby_executing_command === `build`) {
-  delete queueOptions.filter
-  delete queueOptions.priority
-  delete queueOptions.merge
+const pageQueryRun = (queryJob, result) => {
+  // Send event that the page query finished.
+  boundActionCreators.pageQueryRun({
+    path: queryJob.id,
+    componentPath: queryJob.componentPath,
+    isPage: queryJob.isPage,
+  })
 }
 
-const queue = new Queue((plObj, callback) => {
-  const state = store.getState()
-  processing.add(plObj.id)
-
-  return queryRunner(plObj, state.components[plObj.component])
-    .catch(e => console.log(`Error running queryRunner`, e))
-    .then(
-      result => {
-        processing.delete(plObj.id)
-        if (waiting.has(plObj.id)) {
-          queue.push(waiting.get(plObj.id))
-          waiting.delete(plObj.id)
-        }
-
-        // Send event that the page query finished.
-        boundActionCreators.pageQueryRun({
-          path: plObj.id,
-          componentPath: plObj.componentPath,
-          isPage: plObj.isPage,
-        })
-
-        return callback(null, result)
-      },
-      error => callback(error)
-    )
-}, queueOptions)
-
-// HACKY!!! TODO: REMOVE IN NEXT REFACTOR
-// We start paused until we call `runInitialQueries` during bootstrap.
-let isBootstrapping = true
-queue.pause()
-
-emitter.on(`START_QUERY_QUEUE`, () => {
-  isBootstrapping = false
-  queue.resume()
-})
-// END HACKY
-
-// Pause running queries when new nodes are added (processing starts).
-emitter.on(`CREATE_NODE`, () => {
-  queue.pause()
-})
+const makeDaemonOptions = ({ processing, waiting }) => {
+  return {
+    merge: (oldTask, newTask, cb) => {
+      cb(null, newTask)
+    },
+    priority: (job, cb) => {
+      const activePaths = Array.from(websocketManager.activePaths.values())
+      if (job.id && activePaths.includes(job.id)) {
+        cb(null, 10)
+      } else {
+        cb(null, 1)
+      }
+    },
+    // Filter out new query jobs if that query is already running.  When the
+    // query finshes, it checks the waiting map and pushes another job to
+    // make sure all the user changes are captured.
+    filter: (job, cb) => {
+      if (processing.has(job.id)) {
+        waiting.set(job.id, job)
+        cb(`already running`)
+      } else {
+        cb(null, job)
+      }
+    },
+  }
+}
 
-// Resume running queries as soon as the api queue is empty.
-emitter.on(`API_RUNNING_QUEUE_EMPTY`, () => {
-  if (!isBootstrapping) {
-    queue.resume()
+/**
+ * Creates a queue that is optimized for running as a daemon during
```

Comment:
curious about the terminology — a daemon is normally a separate process right?

Reply:
Yeah, I'm not sold on `daemon` as terminology either. I'm trying to communicated that it's a long running process that won't end. Other ideas? `startListener`, `startProcess`?

Reply:
@pieh and I and some others have talked about modelling more of Gatsby's internals after the Actor Model. It's a larger discussion to conclude if we want to do that but "startActor" would be a natural name then.

Reply:
If it is actually an actor, then for sure :). But until then, we should probably stick to something generic so we don't confuse people (and yes, `daemon` is also a bit confusing).

Reply:
If it believes it's an actor it is! Haha.

Service is a nice generic name :-)

Reply:
Yep, I like service. Will change